### ISSUE TITLE
PHPMussel integration in a simple code web application

### ISSUE THREAD

**Issue by gjoivan** [Maikuolan/phpMussel#233]
Hello Caleb I'm writing a code for a uploading a simple file on AWS s3 bucket and I want to perform an antivirus check before it is uploaded. I can't afford to pay for it so I found your solution online -> PHPMussel, and as per the instructions I managed to install it and integrate it to the code with the API example.
I've also created a Scan_Class which has 2 methods in it vt_scan() and vt_report(), one fore sending the file to scan at VirusTotal and the other one for getting the result afterwards, by changing the cURL's URL from "scan" to "report".
Now my question is even though the script returns no problems found or (int-0) the VT API scan isn't complete, I need to wait for 4-5 min than do another API call with "report" in the URL and see the actual API scan results.
How can I modify the script so that the method scan() in Scanner returns the API response (which is always: pending for analysis) and not only "No problems found". I want to see the API response after method scan(), and not need I do var_dum(Response) in the request() method.
Because I think my code will continue after the API call is made and "No problems found is returned", and the VT API scan is not finished, and after that it can turn out that it can be infected
Thanks in advance,
Kind regards

**Comment by Maikuolan** [comment]
Hi gjoivan,

Happy to help. :-)

> by changing the cURL's URL from "scan" to "report".

In order to me to better understand the situation exactly, to clarify, are you referring specifically to this part of the codebase:

https://github.com/phpMussel/Core/blob/37303193d85834fb531589f6b89f62f1aab05627/src/Scanner.php#L2554-L2560
```PHP
                    $VTRequest = $this->Loader->Request->request(
                        'https://www.virustotal.com/vtapi/v2/file/report?apikey=' .
                        urlencode($this->Loader->Configuration['virustotal']['vt_public_api_key']) .
                        '&resource=' . $md5,
                        $VTParams,
                        12
                    );
```

Or to something else entirely (e.g., a cURL request performed from within one of your own classes, etc)?

> Because I think my code will continue after the API call is made and "No problems found is returned", and the VT API scan is not finished, and after that it can turn out that it can be infected

Yeah. Currently, the approach to Virus Total integration that phpMussel uses is to check whether something has already been scanned before at Virus Total by looking up the checksum at the API for whatever's being scanned, and to work with those results from those prior scan attempts, rather than sending the entire file to the API to attempt to scan it anew. The reason for that approach is due to that most webservers will kill the request, and automatically timeout if the time required to complete the request takes longer than ~30, ~60 seconds thereabouts (depending on how whoever controls the webserver has chosen to configure it), plus due to that many people will complain about the scanner being too slow and requiring too much time to complete the scan if it needs to wait the ~3, ~4 minutes thereabouts often needed to finish scanning a file against all engines at the API.

The positives of that approach are that lookups are usually pretty fast and we don't see any complaints from people about the time required to complete the lookups, plus the elimination of potential privacy concerns (e.g., a user tries to upload a file which contains sensitive information, user discovers it was shared with the third-party, Virus Total, which mentions at their homepage about sharing samples onwards with their own partners for the purpose for malware search, improving virus detection and etc, user then complains to webmaster about the file being shared, makes legal threats to the webmaster or similar; to my knowledge, that has never happened before, and isn't particularly relevant to the overall issue at hand, but just mention solely as one of the positives of the approach of looking up only via checksums rather than entire files).

The negatives of that approach are, most notably, as you've mentioned, that a file could potentially be accepted or trusted when it shouldn't be ("No problems found"), particularly so if the specific permutation of the file in question hasn't been encountered before, thus its specific checksum having not been known by the API, causing the API to return results suggesting, at least initially, that the file in question isn't infected, only for it to be discovered as being infected at later, subsequent scans. Sending actual complete samples to the API would, naturally, be the most secure approach, with the highest likelihood of detecting whether a file is actually infected, albeit requiring a little more time.

It should be possible to change the approach, to ensure that actual complete samples are sent, that complete results are waited for, and that phpMussel shows those fresh, complete results accordingly, rather than just results from prior scan attempts. Though, the best approach may depend a little on exactly what expectations there are regarding how phpMussel will interact with your own classes, regarding user experience, etc (e.g., whether we want the request to actually wait for the ~3, ~4 minutes thereabouts required to check the sample against all engines before returning any requests and ending the request, whether we plan to split things up into multiple requests, such as via ajax/ajaj, a sequence of asynchronous requests between an interface and the backend, whether we even need to worry about timeouts at all, due to it maybe being processed via a terminal instead or something like that, or something else entirely, etc).

The exact part in the codebase where the "No problems found" message is generated is located within the `atHit` method, here:

https://github.com/phpMussel/Core/blob/37303193d85834fb531589f6b89f62f1aab05627/src/Loader.php#L1284
```PHP
        /** Fallback for missing text for formatted text. */
        if (!$TextLength) {
            if ($Code === 0) {
                $Text = sprintf(
                    $this->L10N->getString('grammar_exclamation_mark'),
                    sprintf($this->L10N->getString('x_does_not_exist'), $Name)
                );
            } elseif ($Code === 1) {
                $Text = $this->L10N->getString('scan_no_problems_found');
            } else {
                $Text = $this->L10N->getString('data_not_available');
            }
        }
```

Though, it applies more broadly to any kind of scan operation performed by phpMussel, as opposed to something specific to Virus Total, and is some default text filled into the scan results at the end of the scan process for when the scan process doesn't detect anything.

I suspect, where changes will most likely need to be made would *probably* be somewhere around where the Virus Total integration begins proper:

https://github.com/phpMussel/Core/blob/37303193d85834fb531589f6b89f62f1aab05627/src/Scanner.php#L2521
```PHP
        /** Virus Total API integration. */
        if (
            empty($this->Loader->ScanResultsText[$AtInstanceLookupKey]) &&
            !empty($this->Loader->Configuration['virustotal']['vt_public_api_key'])
        ) {
```

But, depending on exactly what expectations there are (mentioned earlier in this reply), there may be other parts of the codebase that need changing, or.. potentially, as long as existing code isn't likely to get in the way, it might be possible to utilise phpMussel's events handler to adjust how it behaves without necessarily needing to touch any of the existing codebase (a more Composer-friendly solution due to less risk of changes being overwritten when updating dependencies, no need to repeat the changes a second time if updating to a newer phpMussel version in the future and so on). If there are some specific parts of the codebase which we're able to identify as potentially getting in the way of what you want to do, depending on how extensive those parts of the codebase are and what they're doing, it might be possible even to add in some additional event hooks at those points in the codebase in order to allow utilising events there for in the future maybe.

**Comment by gjoivan** [comment]
Hey Caleb thanks for the response :) 
This is the class I've created and where I change the scan/report property.
![Screenshot 2023-01-18 at 11 12 03](https://user-images.githubusercontent.com/97735104/213144649-6f8c14b3-5ca0-40b1-998b-b4d4fc3dc185.png)
 And yes that is where I'm changing it: request()- URL
![Screenshot 2023-01-18 at 11 14 04](https://user-images.githubusercontent.com/97735104/213144940-5605f0ef-1329-4a5b-96a2-a76ebf91949f.png)
So depending on needs I'll call either method to scan or get a report on specific fie.
![Screenshot 2023-01-18 at 11 19 32](https://user-images.githubusercontent.com/97735104/213147535-e5bbfb82-60db-47f6-bb1d-21d17f881a8d.png)


I have one more question can I use the API integration example without the API itself, is it giving a proper protection against viruses just by calling the scan() function on a file with the vt_suspicion_level set at 1 or 0 (if no API request is made to VT, and I get "No problems found" is that something I can rely on).
Also is there a way to update the Signatures (how often), isn't it risky to stay outdated.

**Comment by Maikuolan** [comment]
Hm.. I've only just realised that I haven't documented `vt_suspicion_level` properly since v3. I'll need to get that sorted out.

> is it giving a proper protection against viruses just by calling the scan() function on a file

Possibly yes, possibly no, depending on the exact scenario where the scan is occurring, and on exactly how phpMussel has been configured.

For example: `delete_on_sight` instructs phpMussel to delete a file immediately upon detecting that the file is infected with something. Might be useful to switch that on if your application isn't already coded to do something with infected files, to ensure that phpMussel can deal with them (though best to switch it off during testing, to ensure test files aren't deleted by mistake).

In the absence of an application or any such thing coded to respond appropriately to when a file is flagged as either infected or not infected, [phpMussel/Web](https://github.com/phpMussel/Web) could also be useful, too, for protecting normal websites (or might be useful even as reference material for integrating phpMussel into your own application). Though, using it conjunction with an application coded as such, it may potentially get in the way of that application's interface or other related functionality, seeing as it attempts to take responsibility for that itself.

Configured correctly, phpMussel should provide a reasonable degree of protection, and compared to the very limited number of other free, open-source, PHP-based anti-virus solutions available online that I've been able to find over the years, it should do a much better job of it (there are a few, but certainly, there aren't many, and IMHO, while I won't say all of them, most of them, by far, aren't particularly good).

That said, taking in mind that phpMussel is free, that I'm working on it basically by myself (with occasional assistance, input, suggestions, advice, and pull requests from others over the years, which I of course appreciate), and taking in mind that it's a hobbyist project (i.e., no big company behind it), there are always going to limitations and imperfections, and while I'd still suggest it's better to have it than to not have it, I'd be wary of treating it like a "magic bullet".

The main reason why I chose for phpMussel to produce that exact wording ("no problems found"), instead of something like "file isn't infected", "file is safe", "file is clean", etc, is for that exact reason: To convey that nothing bad has been found, but without the unwavering confidence and certainty that the file is absolutely 100% safe (e.g., there might be something which it wasn't able to find, a very new kind of virus, or some kind of threat it hasn't been coded to deal with), because unwavering confidence and certainty can itself be rather dangerous in some circumstances.

> with the vt_suspicion_level set at 1 or 0 (if no API request is made to VT, and I get "No problems found" is that something I can rely on).

So, with `vt_suspicion_level`, if it's set to exactly 1, phpMussel will perform the Virus Total lookup if it detects that the file being scanned is executable, or could contain something executable (e.g., Windows PE files, Linux ELF files, Mach-O files, DOCX files, ZIP files, etc). Files not matching that criteria won't be looked up.

If it's set to something greater than 1, everything will be lookup up, no matter what.

If it's set to 0, lookups will *only* be performed if the file matches against some already existent internal heuristic detection criteria within phpMussel prior to the point in the codebase where lookups would normally be performed (common fingerprints associated with a range of potentially infected files, but which also aren't guarantees of something being infected, thus the extra lookup as a "second opinion"; most files typically won't meet that critieria though).

My bad for not documenting that for v3. I'll document that shortly though.

**Comment by gjoivan** [comment]
Hey Caleb thanks for the response :),
I just want to know if the script is actually performing a virus check just with the scan() method only, and without the API request to VT (example: vt_suspicion_level: 1 or 0), if yes - I can start implementing it in the code and later on see if I can sort out the proper way of using the API (like I've mentioned above, so far it works but I need to call it separately for scan and report, and still haven't figured out how that is going to be managed).
And also is it possible to update the signatures frequently.

**Comment by Maikuolan** [comment]
Sorry for the delayed reply. Bit busier than expected yesterday.

> My bad for not documenting that for v3. I'll document that shortly though.

And done. :-)

> I just want to know if the script is actually performing a virus check just with the scan() method only, and without the API request to VT (example: vt_suspicion_level: 1 or 0)

Yep, it is (assuming the signature files have been installed and phpMussel configured as per desired). :-)

You can also try scanning some of the test files, or try scanning an EICAR file, to confirm whether everything is working as intended if need be.

> And also is it possible to update the signatures frequently.

In previous major versions, I had a self-updater bundled into the front-end, but given that the self-updater was its own thing, and wasn't integrated with Composer at all, it didn't really fit particularly well with the aim of having v3 be Composer oriented, so I ended up scrapping it for v3.

__[phpMussel/SigTool](https://github.com/phpMussel/SigTool)__ could help, if you're okay with downloading ClamAV's signatures database and then processing them for phpMussel's usage via terminal or similar using the script in question. Otherwise, each new v3 release, I've been doing that myself and bundling the resulting prepared signature files into ZIPs at the __[phpMussel/Examples](https://github.com/phpMussel/Examples)__ repository.

I haven't come up with anything quicker or more frequent yet, though I hope to be able to do so at some point.

**Comment by gjoivan** [comment]
Hey Caleb, once again thanks for the support :)

I've managed to implement the script and so far it works with the vt_scan() method I've created. In future I plan to activate the method for getting the report form VT, once the tool is created on the front end I'll just try to connect it.

Also maybe some time in the future I'll bother you again for guidance on updating the signatures :D, or maybe a single command: "composer require phpmussel/sigtool" will do updates, I'll figure out with time.

Kind Regards, and thank you.
